### PR TITLE
Calling user listeners [11588]

### PR DESCRIPTION
**Pull request by IkerLuengo**
Calling the user listeners from the database queues once the data/entity has been added to the database.

### CODE REVIEW

**Comment by MiguelCompany** [comment]
@IkerLuengo This needs a rebase and a linters pass

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Partial review (only changes to `StatisticsBackend.hpp`)

File: src/cpp/Monitor.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+class Topic;
+class DataReader;
+class DataReaderListener;
+
+} // namespace dds
+} // namespace fastdds
+
+namespace statistics_backend {
+
+struct Monitor
```
Comment (MiguelCompany) [inline src/cpp/Monitor.hpp:46]: A bit of doxydoc here, please.

File: src/cpp/database/database_queue.hpp [hunk 2]
```diff
@@ -348,7 +355,20 @@ class DatabaseEntityQueue : public DatabaseQueue<std::shared_ptr<Entity>>
     {
         try
         {
-            database_->insert(front().second);
+            auto id = database_->insert(front().second.entity);
+            if (front().second.entity->kind == EntityKind::HOST ||
+                    front().second.entity->kind == EntityKind::USER ||
+                    front().second.entity->kind == EntityKind::PROCESS ||
+                    front().second.entity->kind == EntityKind::LOCATOR)
```
Comment (MiguelCompany) [inline src/cpp/database/database_queue.hpp:362]: Please use `<constant> == <var>` constructs.

File: test/unittest/StatisticsBackend/CMakeLists.txt [hunk 3]
```diff
@@ -14,10 +14,15 @@
...
+
+# statistics_backend_tests
+##################################################################
 
 add_executable(statistics_backend_tests StatisticsBackendTests.cpp
-    ${PROJECT_SOURCE_DIR}/src/cpp/StatisticsBackend.cpp
-    ${PROJECT_SOURCE_DIR}/src/cpp/types/EntityId.cpp)
+${PROJECT_SOURCE_DIR}/src/cpp/StatisticsBackend.cpp
+${PROJECT_SOURCE_DIR}/src/cpp/exception/Exception.cpp
+${PROJECT_SOURCE_DIR}/src/cpp/types/EntityId.cpp)
```
Comment (MiguelCompany) [inline test/unittest/StatisticsBackend/CMakeLists.txt:25]: Indentation

File: test/unittest/StatisticsBackend/CMakeLists.txt [hunk 4]
```diff
@@ -30,12 +35,89 @@ target_include_directories(statistics_backend_tests PRIVATE
...
             --gtest_filter=statistics_backend_tests.${test_name}:*/statistics_backend_tests.${test_name}/*)
 endforeach()
+
+# calling_user_listeners
+##################################################################
+
+add_executable(calling_user_listeners_tests CallingUserListenersTests.cpp
+${PROJECT_SOURCE_DIR}/src/cpp/StatisticsBackend.cpp
+${PROJECT_SOURCE_DIR}/src/cpp/exception/Exception.cpp
+${PROJECT_SOURCE_DIR}/src/cpp/types/EntityId.cpp)
```
Comment (MiguelCompany) [inline test/unittest/StatisticsBackend/CMakeLists.txt:54]: Indentation

File: test/unittest/StatisticsParticipantListener/CMakeLists.txt [hunk 5]
```diff
@@ -20,8 +20,8 @@ if(GTEST_FOUND AND GMOCK_FOUND)
     find_package(Threads REQUIRED)
 
     add_executable(statistics_participant_listener_tests StatisticsParticipantListenerTests.cpp
-        ${PROJECT_SOURCE_DIR}/src/cpp/exception/Exception.cpp
-        ${PROJECT_SOURCE_DIR}/src/cpp/topic_types/types.cxx
+    ${PROJECT_SOURCE_DIR}/src/cpp/exception/Exception.cpp
+    ${PROJECT_SOURCE_DIR}/src/cpp/topic_types/types.cxx
```
Comment (MiguelCompany) [inline test/unittest/StatisticsParticipantListener/CMakeLists.txt:24]: Indentation

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: include/fastdds-statistics-backend/StatisticsBackend.hpp [hunk 6]
```diff
@@ -428,8 +434,80 @@ class FASTDDS_STATISTICS_BACKEND_DllAPI StatisticsBackend
...
+        UNDISCOVERY,    ///< The entity was undiscovered
+        UPDATE          ///< The entity was updated
+    };
+
+    /**
+     * @brief Notify the user about a new discovered entity
+     *
+     * @param domain_id The domain where the entity was discovered
+     * @param entity_id The entity_id of the discovered entity
+     * @param entity_kind CallbackKind of the discovery event
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply include/fastdds-statistics-backend/StatisticsBackend.hpp:453]: Absolutely

File: include/fastdds-statistics-backend/StatisticsBackend.hpp [hunk 7]
```diff
@@ -428,8 +434,80 @@ class FASTDDS_STATISTICS_BACKEND_DllAPI StatisticsBackend
...
+            DiscoveryStatus discovery_status);
+
+    /**
+     * @brief Notify the user about a new discovered entity
+     *
+     * There is no DiscoveryStatus parameter because physical entities are never undiscovered nor updated
+     *
+     * @param participant_id Entity ID of the participant that discovered the entity.
+     * @param entity_id The entity_id of the discovered entity
+     * @param entity_kind CallbackKind of the discovery event
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply include/fastdds-statistics-backend/StatisticsBackend.hpp:469]: :+1:

File: src/cpp/StatisticsBackend.cpp [hunk 8]
```diff
@@ -261,5 +272,206 @@ std::vector<std::pair<EntityKind, EntityKind>> StatisticsBackend::get_data_suppo
...
 }
 
+void StatisticsBackend::on_data_available(
+        EntityId domain_id,
+        EntityId entity_id,
+        DataKind data_kind)
+{
+    auto monitor = monitors_.find(domain_id);
+    assert(monitor != monitors_.end());
+    if (monitor->second->domain_listener == nullptr ||
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/StatisticsBackend.cpp:281]: :+1:

File: src/cpp/StatisticsBackend.cpp [hunk 9]
```diff
@@ -261,5 +272,206 @@ std::vector<std::pair<EntityKind, EntityKind>> StatisticsBackend::get_data_suppo
...
+            !monitor->second->domain_callback_mask.is_set(CallbackKind::ON_DATA_AVAILABLE))
+    {
+        // No user listener or mask deactivated
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/StatisticsBackend.cpp:284]: :+1:

File: src/cpp/database/database_queue.hpp (hunk 2 above)
Comment (IkerLuengo) [reply src/cpp/database/database_queue.hpp:362]: :+1:

File: test/unittest/StatisticsParticipantListener/CMakeLists.txt (hunk 5 above)
Comment (IkerLuengo) [reply test/unittest/StatisticsParticipantListener/CMakeLists.txt:24]: :+1:

File: src/cpp/Monitor.hpp (hunk 1 above)
Comment (IkerLuengo) [reply src/cpp/Monitor.hpp:46]: :+1:

File: include/fastdds-statistics-backend/StatisticsBackend.hpp [hunk 10]
```diff
@@ -19,30 +19,35 @@
...
 namespace statistics_backend {
 
-namespace database {
+namespace details {
 
-class Database;
+class StatisticsBackendData;
 
-} // namespace database
+} // namespace details
```
Comment (MiguelCompany) [inline include/fastdds-statistics-backend/StatisticsBackend.hpp:39]: We don't need the forward declaration anymore

File: src/cpp/StatisticsBackendData.cpp [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+#include <database/database.hpp>
+
+namespace eprosima {
+namespace statistics_backend {
+
+namespace database {
+
+class DatabaseEntityQueue;
+
+} // namespace database
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.cpp:35]: It seems this forward declaration is not used

File: src/cpp/StatisticsBackendData.cpp [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+void StatisticsBackendData::on_domain_entity_discovery(
+        EntityId domain_id,
+        EntityId entity_id,
+        EntityKind entity_kind,
+        DiscoveryStatus discovery_status)
+{
+    auto monitor = monitors_.find(domain_id);
+    assert(monitor != monitors_.end());
+    if (monitor->second->domain_listener == nullptr)
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.cpp:71]:
```suggestion
    if (nullptr == monitor->second->domain_listener)
```

File: src/cpp/StatisticsBackendData.cpp [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+    {
+        case EntityKind::PARTICIPANT:
+        {
+            if (!monitor->second->domain_callback_mask.is_set(CallbackKind::ON_PARTICIPANT_DISCOVERY))
+            {
+                // mask deactivated
+                return;
+            }
+
+            if (discovery_status == DISCOVERY)
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.cpp:87]: Please use `<constant> == <variable>` constructs (applies to the whole file)

File: src/cpp/StatisticsBackendData.cpp [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+            {
+                monitor->second->participant_status_.on_instance_discovered();
+            }
+            else if (discovery_status == UNDISCOVERY)
+            {
+                monitor->second->participant_status_.on_instance_undiscovered();
+            }
+            monitor->second->domain_listener->on_participant_discovery(domain_id, entity_id,
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.cpp:95]: Should we avoid calling the listener when the status is neither `DISCOVERY` nor `UNDISCOVERY`?

File: src/cpp/StatisticsBackendData.cpp [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+                    monitor->second->participant_status_);
+            monitor->second->participant_status_.on_status_read();
+            break;
+        }
+        case EntityKind::TOPIC:
+        {
+            if (!monitor->second->domain_callback_mask.is_set(CallbackKind::ON_TOPIC_DISCOVERY))
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.cpp:102]: We have almost the same code four times here. Could a common method be used?

File: src/cpp/StatisticsBackendData.cpp [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+                monitor->second->datareader_status_.on_instance_undiscovered();
+            }
+            monitor->second->domain_listener->on_datareader_discovery(domain_id, entity_id,
+                    monitor->second->datareader_status_);
+            monitor->second->datareader_status_.on_status_read();
+            break;
+        }
+        default:
+        {
+            throw Error("wrong entity_kind");
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.cpp:164]: Should this be an assertion instead?

File: src/cpp/StatisticsBackendData.cpp [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+        }
+    }
+}
+
+void StatisticsBackendData::on_physical_entity_discovery(
+        EntityId participant_id,
+        EntityId entity_id,
+        EntityKind entity_kind)
+{
+    if (physical_listener_ == nullptr)
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.cpp:174]:
```suggestion
    if (nullptr == physical_listener_)
```

File: src/cpp/StatisticsBackendData.cpp [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+    {
+        // No user listener
+        return;
+    }
+
+    switch (entity_kind)
+    {
+        case EntityKind::HOST:
+        {
+            if (!physical_callback_mask_.is_set(CallbackKind::ON_HOST_DISCOVERY))
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.cpp:184]: Again, we have similar code four times. Could a common method be used?

File: src/cpp/StatisticsBackendData.cpp [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+            }
+
+            locator_status_.on_instance_discovered();
+            physical_listener_->on_locator_discovery(participant_id, entity_id, locator_status_);
+            locator_status_.on_status_read();
+            break;
+        }
+        default:
+        {
+            throw Error("wrong entity_kind");
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.cpp:236]: Assertion?

File: src/cpp/StatisticsBackendData.hpp [hunk 20]
```diff
@@ -0,0 +1,151 @@
...
+#include "Monitor.hpp"
+
+namespace eprosima {
+namespace statistics_backend {
+
+namespace database {
+
+class DatabaseEntityQueue;
+
+} // namespace database
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.hpp:41]: It seems this forward declaration is not used in the file.

File: src/cpp/StatisticsBackendData.hpp [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+
+namespace details {
+
+struct Monitor;
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackendData.hpp:45]: We are including `Monitor.hpp`, so this forward declaration is redundant.

File: test/unittest/DatabaseQueue/DatabaseQueueTests.cpp [hunk 22]
```diff
@@ -22,11 +22,14 @@
...
 using namespace eprosima::statistics_backend::database;
-using EntityId = eprosima::statistics_backend::EntityId;
-using DataKind = eprosima::statistics_backend::DataKind;
-using EntityKind = eprosima::statistics_backend::EntityKind;
-using Timestamp = eprosima::statistics_backend::Timestamp;
+
+using EntityId = EntityId;
+using DataKind = DataKind;
+using EntityKind = EntityKind;
+using Timestamp = Timestamp;
```
Comment (MiguelCompany) [inline test/unittest/DatabaseQueue/DatabaseQueueTests.cpp:32]: Unnecessary?

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/StatisticsBackendData.cpp (hunk 14 above)
Comment (IkerLuengo) [reply src/cpp/StatisticsBackendData.cpp:95]: We decided to inform the application also in the case of modified QoS. The application can get the modified status from the QoS stored in the database, once it knows the modified entity ID.

File: src/cpp/StatisticsBackend.cpp [hunk 23]
```diff
@@ -19,23 +19,20 @@
...
 
-using namespace eprosima::statistics_backend::database;
-
-Database* StatisticsBackend::database_ = new database::Database();
 
 void StatisticsBackend::set_physical_listener(
         PhysicalListener* listener,
         CallbackMask callback_mask,
-        DataKindMask data_mask)
+        DataKindMask /*data_mask*/)
```
Comment (MiguelCompany) [inline src/cpp/StatisticsBackend.cpp:32]: Should the physical listener be called for other kind of data?

**Review by MiguelCompany — APPROVED** [review]
LGTM. Just one minor concern to check we are on the same page

File: src/cpp/StatisticsBackendData.cpp [hunk 24]
```diff
@@ -38,20 +38,81 @@ void StatisticsBackendData::on_data_available(
...
+        return false;
+    }
+
+    if (!monitor->domain_callback_mask.is_set(callback_kind))
+    {
+        // mask deactivated
+        return false;
+    }
+
+    if (data_kind != DataKind::INVALID && !monitor->data_mask.is_set(data_kind))
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/StatisticsBackendData.cpp:71]: Yes. Added some comments (and doxydoc ref in the method) clarifying this

File: src/cpp/StatisticsBackendData.cpp [hunk 25]
```diff
@@ -38,20 +38,81 @@ void StatisticsBackendData::on_data_available(
...
+        return false;
+    }
+
+    if (!physical_callback_mask_.is_set(callback_kind))
+    {
+        // mask deactivated
+        return false;
+    }
+
+    if (data_kind != DataKind::INVALID && !physical_data_mask_.is_set(data_kind))
```
[the remark this answers is not in the document]
Comment (IkerLuengo) [reply src/cpp/StatisticsBackendData.cpp:98]: same as before

**Review by MiguelCompany — APPROVED** [review]

**Pull request merged**